src/Mvc/Mvc.TagHelpers/src/PersistComponentStateTagHelper.cs

```diff
@@ -0,0 +1,89 @@
+        private const string PersistenceModeName = "persist-mode";
+        private PersistenceMode? _persistenceMode;
+
+        /// <summary>
+        /// Gets or sets the <see cref="Rendering.ViewContext"/> for the current request.
+        /// </summary>
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        /// <summary>
+        /// Gets or sets the <see cref="PersistenceMode"/> for the state to persist.
+        /// </summary>
+        [HtmlAttributeName(PersistenceModeName)]
+        public PersistenceMode? PersistenceMode
+        {
+            get => _persistenceMode;
+            set => _persistenceMode = value;
+        }
+
+        /// <inheritdoc />
+        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (output == null)
+            {
+                throw new ArgumentNullException(nameof(output));
+            }
+
+            var services = ViewContext.HttpContext.RequestServices;
+            var manager = services.GetRequiredService<ComponentApplicationLifetime>();
+            var renderer = services.GetRequiredService<HtmlRenderer>();
+            var store = PersistenceMode switch
+            {
+                null => ComponentRenderer.GetPersistStateRenderMode(ViewContext) switch
+                {
+                    InvokedRenderModes.Mode.None =>
+                        null,
+                    InvokedRenderModes.Mode.WebAssembly =>
+                        new PrerenderComponentApplicationStore(),
+                    InvokedRenderModes.Mode.Server =>
+                        new ProtectedPrerenderComponentApplicationStore(services.GetRequiredService<IDataProtectionProvider>()),
+                    InvokedRenderModes.Mode.ServerAndWebAssembly =>
+                        throw new InvalidOperationException("The persistence mode could not be determine based on the rendered components. Specify the PersistenceMode explicitly."),
+                    _ => throw new InvalidOperationException("Invalid InvokedRenderMode.")
+                },
+                TagHelpers.PersistenceMode.Server =>
+                    new ProtectedPrerenderComponentApplicationStore(services.GetRequiredService<IDataProtectionProvider>()),
+                TagHelpers.PersistenceMode.WebAssembly =>
+                    new PrerenderComponentApplicationStore(),
+                _ =>
+                    throw new InvalidOperationException("Invalid persistence mode.")
+            };
+
+            output.TagName = null;
+            if (store != null)
+            {
+                await manager.PersistStateAsync(store, renderer);
+                output.Content.SetHtmlContent(new PersistedStateContent(store.PersistedState));
```

Comment:
You could use a `HtmlContentBuilder` instead of creating a new type.

Reply:
I wasn't aware of HtmlContentBuilder, I followed the pattern we have with the serializers.

Reply:
I've changed this one and I'll revisit others in the future